### ISSUE TITLE
Bleadperl v5.13.8-391-g7fe50b8 breaks MNOONING/Net-Daemon/Net-Daemon-0.43.tar.gz

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11121]
Migrated from [rt.perl.org#83646](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=83646) (status was 'resolved')

Searchable as RT83646$

**Comment by p5pRT** [comment]
### From @andk

git bisect&#8203;:


---
&nbsp;  7fe50b8b8a4dc38fc341e3b403545aaca937f50e is the first bad commit
&nbsp;  commit 7fe50b8b8a4dc38fc341e3b403545aaca937f50e
&nbsp;  Author&#8203;: Leon Timmermans \<fawaka@&#8203;gmail\.com>
&nbsp;  Date&#8203;:   Tue Jan 18 16&#8203;:40&#8203;:07 2011 \+0100

&nbsp;  Also unblock signal handlers throwing an exception

&nbsp;  Also handle and test the edge case of a signal handler throwing an
&nbsp;  exception

symptoms&#8203;:


---
&nbsp;  The test forkm\.t now hangs\. Note that N&#8203;:D tests with threaded perls
&nbsp;  are broken for several years but there's a patch in RT ticket \#55512
&nbsp;  and this is a completely separate issue\.

perl \-V&#8203;:


---
&nbsp;  Summary of my perl5 \(revision 5 version 13 subversion 8\) configuration&#8203;:
&nbsp;    Commit id&#8203;: 7fe50b8b8a4dc38fc341e3b403545aaca937f50e
&nbsp;    Platform&#8203;:
&nbsp;      osname=linux\, osvers=2\.6\.32\-5\-xen\-amd64\, archname=x86\_64\-linux\-ld
&nbsp;      uname='linux k81 2\.6\.32\-5\-xen\-amd64 \#1 smp wed jan 12 05&#8203;:46&#8203;:49 utc 2011 x86\_64 gnulinux '
&nbsp;      config\_args='\-Dprefix=/home/src/perl/repoperls/installed\-perls/perl/v5\.13\.8\-391\-g7fe50b8 \-Dinstallusrbinperl=n \-Uversiononly \-Dusedevel \-des \-Ui\_db \-Uuseithreads \-Duselongdouble'
&nbsp;      hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;      useithreads=undef\, usemultiplicity=undef
&nbsp;      useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;      use64bitint=define\, use64bitall=define\, uselongdouble=define
&nbsp;      usemymalloc=n\, bincompat5005=undef
&nbsp;    Compiler&#8203;:
&nbsp;      cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;      optimize='\-O2'\,
&nbsp;      cppflags='\-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;      ccversion=''\, gccversion='4\.4\.5'\, gccosandvers=''
&nbsp;      intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;      d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;      ivtype='long'\, ivsize=8\, nvtype='long double'\, nvsize=16\, Off\_t='off\_t'\, lseeksize=8
&nbsp;      alignbytes=16\, prototype=define
&nbsp;    Linker and Libraries&#8203;:
&nbsp;      ld='cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;      libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
&nbsp;      libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc \-lgdbm\_compat
&nbsp;      perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;      libc=/lib/libc\-2\.11\.2\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;      gnulibc\_version='2\.11\.2'
&nbsp;    Dynamic Linking&#8203;:
&nbsp;      dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;      cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-L/usr/local/lib \-fstack\-protector'


&nbsp;  Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;    Compile\-time options&#8203;: PERL\_DONT\_CREATE\_GVSV PERL\_MALLOC\_WRAP PERL\_USE\_DEVEL
&nbsp;                          USE\_64\_BIT\_ALL USE\_64\_BIT\_INT USE\_LARGE\_FILES
&nbsp;                          USE\_LONG\_DOUBLE USE\_PERLIO USE\_PERL\_ATOF
&nbsp;    Built under linux
&nbsp;    Compiled at Feb  6 2011 22&#8203;:13&#8203;:21
&nbsp;    @&#8203;INC&#8203;:
&nbsp;      /home/src/perl/repoperls/installed\-perls/perl/v5\.13\.8\-391\-g7fe50b8/lib/site\_perl/5\.13\.8/x86\_64\-linux\-ld
&nbsp;      /home/src/perl/repoperls/installed\-perls/perl/v5\.13\.8\-391\-g7fe50b8/lib/site\_perl/5\.13\.8
&nbsp;      /home/src/perl/repoperls/installed\-perls/perl/v5\.13\.8\-391\-g7fe50b8/lib/5\.13\.8/x86\_64\-linux\-ld
&nbsp;      /home/src/perl/repoperls/installed\-perls/perl/v5\.13\.8\-391\-g7fe50b8/lib/5\.13\.8
&nbsp;      \.


\-\- 
andreas

**Comment by p5pRT** [comment]
### From @Leont

On Tue\, Feb 8\, 2011 at 7&#8203;:01 AM\, Andreas J\. Koenig via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  \(Andreas J\. Koenig\)
> \# Please include the string&#8203;:  \[perl \#83646\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=83646 >

That module is a convoluted mess that already fails its unit tests
half of the time\[1\]\, even statistical analysis\[2\] doesn't really
clarify things  all that much\. I'm tempted to blame the module's
fickleness\.

Leon

\[1\] http&#8203;://matrix\.cpantesters\.org/?dist=Net\-Daemon\+0\.43
\[2\] http&#8203;://analysis\.cpantesters\.org/solved?distv=Net\-Daemon\-0\.43

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Mon\, 14 Feb 2011 04&#8203;:00&#8203;:16 \-0800\, "Leon Timmermans via RT" \<perlbug\-followup@&#8203;perl\.org> said&#8203;:

&nbsp;  > On Tue\, Feb 8\, 2011 at 7&#8203;:01 AM\, Andreas J\. Koenig via RT
&nbsp;  > \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
 >> \# New Ticket Created by  \(Andreas J\. Koenig\)
 >> \# Please include the string&#8203;:  \[perl \#83646\]
 >> \# in the subject line of all future correspondence about this issue\.
 >> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=83646 >

&nbsp;  > That module is a convoluted mess that already fails its unit tests
&nbsp;  > half of the time\[1\]\, even statistical analysis\[2\] doesn't really
&nbsp;  > clarify things  all that much\. I'm tempted to blame the module's
&nbsp;  > fickleness\.

What you say is unfair\.

That the matrix at \[1\] can't show you much is due to the fact that \(I'm
citing from this ticket\)&#8203;:

&nbsp;    N&#8203;:D tests with threaded perls are broken for several years but
&nbsp;    there's a patch in RT ticket \#55512 and this is a completely
&nbsp;    separate issue\.

The statistical analysis at \[2\] cannot deal with two issues at once\.
That's why I write these tickets before too much of CPAN burns\.

&nbsp;  > \[1\] http&#8203;://matrix\.cpantesters\.org/?dist=Net\-Daemon\+0\.43
&nbsp;  > \[2\] http&#8203;://analysis\.cpantesters\.org/solved?distv=Net\-Daemon\-0\.43

As a matter of fact\, statistical analysis before v5\.13\.8\-391\-g7fe50b8
was giving quite a clear result that \-Dusethreads does not work\. And
combine this with the fact that the test forkm\.t starts to hang on non
threaded perls exactly since v5\.13\.8\-391\-g7fe50b8\.

You may argue about the quality of the module but what you're saying
here is ignoring the facts\.

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @Leont

On Mon\, Feb 14\, 2011 at 8&#8203;:50 PM\, Andreas J\. Koenig
\<andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de> wrote&#8203;:
> What you say is unfair\.

With «I'm tempted to blame the module's fickleness\.» I didn't mean
«screw him»\, but more something like «the code is just plain wrong in
so many ways I'm surprised it works at all»\.

Anyway\, it seems I've figured this out\. The SIGCHLD handler calls exit
when all clients have died\, but forgets about the server\. It just
happened to work because the program never received any SIGCHLD after
all clients were waited for\. Previously\, the SIGCHLD was never
unblocked since the handler never returns\, now it is unblocked as part
of stack\-unwinding\. Unfortunately\, the SIGCHLD handler was really
badly written \(if not outright crazy\) and an unexpected SIGCHLD
\(probably from a child\) is enough to get it into a blocking wait \+
infinite loop because it insists on receiving signals from dead
clients while there's only a living server\.

The module is at fault\. If he had read perlipc before writing signal
handlers he would have known how to write a signal handler\. See the
patch for a simple solution\.

> As a matter of fact\, statistical analysis before v5\.13\.8\-391\-g7fe50b8
> was giving quite a clear result that \-Dusethreads does not work\.

Not quite\, \-Dusethreads does work in a minority of cases \(and the
other way around it sometimes doesn't work on threadless perls\, though
I guess that's noise\)\. Hence the R^2 is 0\.552&#8203;: significant but not the
entire story\. There are other factors playing an similarly big role in
this\. I have no idea what they are though\. The module is fickly IMO\.

Leon

**Comment by p5pRT** [comment]
### From @Leont

<details><summary>waitpid.patch</summary>

```diff
diff --git a/t/forkm.t b/t/forkm.t
index 8dde319..03e4547 100644
--- a/t/forkm.t
+++ b/t/forkm.t
@@ -8,6 +8,7 @@ use Config ();
 use Net::Daemon::Test ();
 use Fcntl ();
 use Config ();
+use POSIX qw/WNOHANG/;
 
 
 my $debug = 0;
@@ -130,7 +131,8 @@ my %childs;
 sub CatchChild {
     &log("CatchChild: ->");
     for(;;) {
-	my $pid = wait;
+	my $pid = waitpid -1, WNOHANG;
+	last if $pid <= 0;
 	if ($pid > 0) {
 	    &log("CatchChild: $pid");
 	    if (exists $childs{$pid}) {

```
</details>

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Tue\, 15 Feb 2011 02&#8203;:48&#8203;:36 \+0100\, Leon Timmermans \<fawaka@&#8203;gmail\.com> said&#8203;:

&nbsp;  > On Mon\, Feb 14\, 2011 at 8&#8203;:50 PM\, Andreas J\. Koenig
&nbsp;  > \<andreas\.koenig\.7os6VVqR@&#8203;franz\.ak\.mind\.de> wrote&#8203;:
 >> What you say is unfair\.

&nbsp;  > With «I'm tempted to blame the module's fickleness\.» I didn't mean
&nbsp;  > «screw him»\, but more something like «the code is just plain wrong in
&nbsp;  > so many ways I'm surprised it works at all»\.

OK\, you did not say that\.

&nbsp;  > Anyway\, it seems I've figured this out\. The SIGCHLD handler calls exit
&nbsp;  > when all clients have died\, but forgets about the server\. It just
&nbsp;  > happened to work because the program never received any SIGCHLD after
&nbsp;  > all clients were waited for\. Previously\, the SIGCHLD was never
&nbsp;  > unblocked since the handler never returns\, now it is unblocked as part
&nbsp;  > of stack\-unwinding\. Unfortunately\, the SIGCHLD handler was really
&nbsp;  > badly written \(if not outright crazy\) and an unexpected SIGCHLD
&nbsp;  > \(probably from a child\) is enough to get it into a blocking wait \+
&nbsp;  > infinite loop because it insists on receiving signals from dead
&nbsp;  > clients while there's only a living server\.

&nbsp;  > The module is at fault\. If he had read perlipc before writing signal
&nbsp;  > handlers he would have known how to write a signal handler\. See the
&nbsp;  > patch for a simple solution\.

Thank you for figuring that out\.

 >> As a matter of fact\, statistical analysis before v5\.13\.8\-391\-g7fe50b8
 >> was giving quite a clear result that \-Dusethreads does not work\.

&nbsp;  > Not quite\, \-Dusethreads does work in a minority of cases \(and the
&nbsp;  > other way around it sometimes doesn't work on threadless perls\, though
&nbsp;  > I guess that's noise\)\. Hence the R^2 is 0\.552&#8203;: significant but not the
&nbsp;  > entire story\. There are other factors playing an similarly big role in
&nbsp;  > this\. I have no idea what they are though\. The module is fickly IMO\.

I admit it's not obvious which factors influence the results but looking
a bit closer I find that in the 5\.8\.\[1\-8\] time Net&#8203;::Daemon was mostly
working under \-Duseithreads\. It started to fail drastically somewhere in
the 5\.9\.5 era with 5\.11 and 5\.13 \*reliably\* failing under \-Duseithreads\.
I understand that it is tempting to blame the module but it is as
tempting to blame perl because perl has changed and Net\-Daemon not\.

The real pity is that the maintainer is AWOL\.

\-\- 
andreas

**Comment by p5pRT** [comment]
### From m.nooning@comcast.net

Thanks for the patch\.  I just uploaded Net\-Daemon\-0\.48\.tar\.gz to CPAN\. 
It should be visible by tomorrow\.  I could only test it with Perl 5\.12
on Windows and Fedora

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
